src/reliable_text.c

```diff
@@ -0,0 +1,444 @@
+}
+
+static char reliable_text_freedv_callback_tx(void *state)
+{
+    reliable_text_impl_t* obj = (reliable_text_impl_t*)state;
+    
+    char ret = obj->tx_text[obj->tx_text_index];
+    obj->tx_text_index = (obj->tx_text_index + 1) % (obj->tx_text_length);
+    
+    //fprintf(stderr, "sent char: %d\n", ret);
+    return ret;
+}
+
+reliable_text_t reliable_text_create()
+{
+    reliable_text_impl_t* ret = calloc(sizeof(reliable_text_impl_t), 1);
+    
+    // Load LDPC code into memory.
+    int code_index = ldpc_codes_find("HRA_56_56");
+    memcpy(&ret->ldpc, &ldpc_codes[code_index], sizeof(struct LDPC));
+    
+    return (reliable_text_t)ret;
+}
+
+void reliable_text_destroy(reliable_text_t ptr)
+{
+    free(ptr);
+}
+
+void reliable_text_reset(reliable_text_t ptr)
+{
+    reliable_text_impl_t* impl = (reliable_text_impl_t*)ptr;
+    impl->bit_index = 0;
+    impl->sym_index = 0;
+    impl->uw_bits_valid = 0;
+    impl->uw_value = 0;
+    impl->has_successfully_decoded = 0;
+    memset(&impl->inbound_pending_syms, 0, sizeof(complex float)*LDPC_TOTAL_SIZE_BITS/2);
+    memset(&impl->inbound_pending_amps, 0, sizeof(float)*LDPC_TOTAL_SIZE_BITS/2);
+}
+
+void reliable_text_set_string(reliable_text_t ptr, const char* str, int strlength)
+{
+    reliable_text_impl_t* impl = (reliable_text_impl_t*)ptr;
+    
+    char tmp[RELIABLE_TEXT_MAX_RAW_LENGTH];
+    
+    convert_callsign_to_ota_string_(str, &tmp[RELIABLE_TEXT_CRC_LENGTH], strlength);
+    
+    int raw_length = strlen(&tmp[RELIABLE_TEXT_CRC_LENGTH]) + RELIABLE_TEXT_CRC_LENGTH;
+    impl->tx_text_length = RELIABLE_TEXT_UW_LENGTH_BITS + LDPC_TOTAL_SIZE_BITS;
+    impl->tx_text_index = 0;
+    unsigned char crc = calculateCRC8_(&tmp[RELIABLE_TEXT_CRC_LENGTH], raw_length - RELIABLE_TEXT_CRC_LENGTH);
+    tmp[0] = crc;
+
+    // Encode block of text using LDPC(112,56).
+    unsigned char ibits[LDPC_TOTAL_SIZE_BITS / 2];
+    unsigned char pbits[LDPC_TOTAL_SIZE_BITS / 2];
+    memset(ibits, 0, LDPC_TOTAL_SIZE_BITS / 2);
+    memset(pbits, 0, LDPC_TOTAL_SIZE_BITS / 2);
+    for (int index = 0; index < 8; index++)
+    {
+        if (tmp[0] & (1 << index)) ibits[index] = 1;
+    }
+
+    // Pack 6 bit characters into single LDPC block.
+    for (int ibitsBitIndex = 8; ibitsBitIndex < (LDPC_TOTAL_SIZE_BITS / 2); ibitsBitIndex++)
+    {
+        int bitsFromCrc = ibitsBitIndex - 8;
+        unsigned int byte = tmp[RELIABLE_TEXT_CRC_LENGTH + bitsFromCrc / 6];
+        unsigned int bitToCheck = bitsFromCrc % 6;
+        //fprintf(stderr, "bit index: %d, byte: %x, bit to check: %d, result: %d\n", ibitsBitIndex, byte, bitToCheck, (byte & (1 << bitToCheck)) != 0);
+        
+        if (byte & (1 << bitToCheck))
+        {
+            ibits[ibitsBitIndex] = 1;
+        }
+    }
+    
+    encode(&impl->ldpc, ibits, pbits);  
+    
+    // Split LDPC encoded bits into individual bits, with the first RELIABLE_TEXT_UW_LENGTH_BITS being UW.
+    memset(impl->tx_text, 1, RELIABLE_TEXT_UW_LENGTH_BITS);
```

Comment:
OK so an all 1's UW?  We usually avoid that as but I think it's OK in this case.

Reply:
The reason we avoid it is occasionally we get edge cases (e.g sync on noise) that lead to the same QPSK symbols being decoded.  The reason I don't think it's a concern here is that the OFDM modem already has solid sync so the data your algorithms get fed won't be too crazy.